impeller/entity/shaders/utils.sl.h

```diff
@@ -0,0 +1,12 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+vec4 ImpellerTexture(sampler2D texture_sampler,
+                     vec2 coords,
+                     float y_coord_scale) {
+  if (y_coord_scale < 0.0) {
+    coords.y = 1.0 - coords.y;
+  }
```

Comment:
I've been under the impresion we want to avoid branches in shader code. Would it be better to try to branch in the CPU code? Maybe I'm missing where this wouldn't work in some other shader because of multiple textures being used or something...

Reply:
This looks like the unity design: https://docs.unity3d.com/540/Documentation/Manual/SL-PlatformDifferences.html

IIRC, as long as `y_coord_scale` comes from a uniform, this shouldn't cause multi execution. The problem comes in when branching on non-uniforms, which will sometimes cause the program to run multiple times on the same fragment.

Reply:
> problem comes in when branching on non-uniforms

That was my understanding as well. But, I haven't profiled the generated shaders.

> it should get inlined at an early stage in GL drivers.

For Metal, the signature is:
```
static inline __attribute__((always_inline))
float4 ImpellerTexture(thread const texture2d<float> texture_sampler, thread const sampler texture_samplerSmplr, thread float2& coords, thread const float& y_coord_scale)
```

Reply:
We should probably have all this info in a comment on the line with the branch.

Reply:
IMO, we should have a separate doc for writing performant shaders and how to profile shaders. I have created an issue for this https://github.com/flutter/flutter/issues/105062. I have already added instructions for debugging shaders here https://github.com/flutter/engine/blob/main/impeller/docs/read_frame_captures.md#debugging-a-shader but another section specifically on profiling makes sense too.